### PR TITLE
Add permentropy and its tests

### LINKED ISSUE #18: Permutation entropy
Hi, I implemented the permutation entropy ([Bandt & Pompe, 2002](http://doi.org/10.1103/PhysRevLett.88.174102)).  Do you want add it to ChaosTools.jl?  Let me know if you want a PR.  Here is a demo: https://gist.github.com/tkf/31cb31564dc4f61c085d013656f1a9e9

```julia
function permutation_entropy(
        time_series::Array{T, 1}, ord::O, sampling_interval::S = 1;
        base = Base.e) where {T, O <: Integer, S <: Integer}
    n = length(time_series)
    perms = hcat(Combinatorics.permutations(O(1):ord)...)
    nperm = size(perms)[2]
    ptype = collect(O(1):ord)
    count = zeros(UInt64, nperm)

    for t in 1:n-sampling_interval * ord
        sample = time_series[t:sampling_interval:t+sampling_interval * (ord-1)]
        sortperm!(ptype, sample)
        for i in 1:nperm
            if perms[:, i] == ptype
                count[i] += 1
                break
            end
        end
    end

    nonzero = collect(c for c in count if c != 0)
    p = nonzero / sum(nonzero)
    return - sum(p .* log.(base, p))
end
```

I think this is not an optimal implementation though.  I just wrote it as a quick exercise a while ago.  Besides a few Julia related improvements (like using `@views`), I think the inner most loop that searches the permutation can be improved, using the fact that the permutation counter `perms` is sorted and maybe also by using better data structure.

Also, I'm not a data analysis guy so I don't know if this method is relevant these days.

Anyway, what do you think?  It's good to add it?

### PR DESCRIPTION
**Pull request by tkf**
closes #18 

Here is a demo: https://gist.github.com/tkf/31cb31564dc4f61c085d013656f1a9e9

### CODE REVIEW

**Comment by Datseris** [comment]
Looking good!!! Before we go into performance discussions, may I suggest a name change to `permentropy`, to have a more similar feel to `genentropy` ?

**Comment by Datseris** [comment]
This is looking really nice, I can merge right away if tests pass.

Let's see this part now:


```
+    nonzero = [c for c in count if c != 0]
+
+    p = nonzero ./ sum(nonzero)
+    return  sum(p .* log.(base, p))
```
If you do `scale = sum(c)` and then do return sum( (p/scale) * log(p/scale) for p in c if p != 0)` wouldn't it be a lot faster?

I think `p .* log.(base, p)` already allocates, and of course its clear that `nonzero = [c for c in count if c != 0]` allocates as well,

File: src/dimensions/entropies.jl [hunk 1]
```diff
@@ -112,3 +114,56 @@ shannon(args...) = genentropy(1, args...)
...
+        count[i] += 1
+    end
+
+    # To compute `p log(p)` correctly for `p = 0`, we first discard
+    # cases with zero occurrence.  They don't contribute to the final
+    # sum hence to the entropy:
+    nonzero = [c for c in count if c != 0]
+
+    p = nonzero ./ sum(nonzero)
+    return - sum(p .* log.(base, p))
```
Comment (Datseris) [inline src/dimensions/entropies.jl]: Will it not be much faster if one does
```
scale = sum(c)
return -sum( (p/scale)*log(p/scale) for p in c if p != 0)
```
? This version does no allocations, the original does 3 allocations, once when defining `nonzero`, once when defining `p` and once when doing `p .* log.(base, p)` (this also allocates a new array)

File: src/dimensions/entropies.jl [hunk 2]
```diff
@@ -112,3 +114,56 @@ shannon(args...) = genentropy(1, args...)
...
+sliding windows between `t0` and `t1 = t0 + interval * (order - 1)`.
+
+## References
+
+[1] : Bandt, C., & Pompe, B., Phys. Rev. Lett. **88** (17), pp 174102 (2002)
+"""
+function permentropy(
+        time_series::AbstractArray{T, 1}, order::UInt8,
+        interval::Integer = 1;
+        base=Base.e) where {T}
```
Comment (Datseris) [inline src/dimensions/entropies.jl]: Having the "base" option is cool of course, but the thing is that all other entropies assume base `e`. Maybe we should add a keyword to all of them? 

(Or just use base-e everywhere, which is my suggestion)

**Comment by tkf** [comment]
Yea it allocates memory but it executed only once.  For long time series I don't think it matters.

`perms = map(PermType, permutations(1:order))` likely allocates much more memory and doing some optimization for `sum(p log(p))` doesn't buy us match, unless we improve `perms`.  I wouldn't worry about it if I were you.

**Comment by tkf** [comment]
Regarding further (memory/CPU) performance improvement around `perms` representation, probably using some kind of tree structure would help.  Maybe this?:
https://www.sciencedirect.com/science/article/pii/0898122194001782

But I'd say this implementation is OK for some moderate `order`.

**Comment by Datseris** [comment]
The Julia people are super good on that stuff, if you care about it a lot you can post a question on Discourse and people will immediately help. I am sure there are special structures that can add a lot of performance.

At the moment it is not crucial for me and I can merge this and then open an issue about its performance.

What about the base thing?

File: src/dimensions/entropies.jl (hunk 2 above)
Comment (tkf) [reply src/dimensions/entropies.jl]: I think `base=2` is also very popular.  In fact, `base=2` is used in the original permutation entropy paper.  I'd like to have base option everywhere.  Or something like `genentropy2`.

**Comment by tkf** [comment]
I commented on base things as the review comment. Personally, I'd like to keep it and have base option everywhere.

**Comment by Datseris** [comment]
> I commented on base things as the review comment. 

Can't find it! Regardless, you are right. Seems like the best option. I will do the change of including base on this PR, is that okay? I will also add my performance benefit, because even if it is small it still exists and does not reduce code clarity.

**Comment by tkf** [comment]
I just wrote:

> I think base=2 is also very popular. In fact, base=2 is used in the original permutation entropy paper. I'd like to have base option everywhere. Or something like genentropy2.

**Comment by tkf** [comment]
Re: performance

I like the saying "If you don't benchmark, you don't care the performance."  We are not doing any benchmarks here so talking about it is kind of nonsense.

OK. Not fair. I was the first one to bringing that up. But I knew that computational complexity of the innner most loop of the original code was `O(N)` where `N=order!` which was ridiculous.  I just switched to `searchsortedfirst` (which uses the binary search) since that would give us `O(log N)`.

**Comment by tkf** [comment]
Probably setting up benchmarks using PkgBenchmark.jl would help us doing further performance improvement.

**Comment by Datseris** [comment]
> Probably setting up benchmarks using PkgBenchmark.jl would help us doing further performance improvement.

Yeah sounds like an excellent plan. So far I have developed myself and benchmarked on the same computer with e.g. `@btime` but now is the time for upgrade.

File: src/dimensions/entropies.jl [hunk 3]
```diff
@@ -112,3 +114,56 @@ shannon(args...) = genentropy(1, args...)
...
+    permutation_entropy(x::AbstractVector, order [, interval=1]; base=e)
+
+Compute the permutation entropy (Bandt & Pompe, 2002) of given `order`
+from the `x` timeseries.  Optionally, `interval` can be specified to
+use `time_series[t0:interval:t1]` when calculating permutation of the
+sliding windows between `t0` and `t1 = t0 + interval * (order - 1)`.
+
+## References
+
+[1] : Bandt, C., & Pompe, B., Phys. Rev. Lett. **88** (17), pp 174102 (2002)
```
[the remark this answers is not in the document]
Comment (Datseris) [reply src/dimensions/entropies.jl]: Daaaaamn how did I not consider to that everywhere!

**Comment by Datseris** [comment]
Alright, I added base to `genentropy`. Assuming I did not have 10s of typos and tests pass, this is merged as-is. A separate issue will be created for its performance.

**Comment by tkf** [comment]
Thanks!

BTW, it looks like Travis is failing

**Comment by Datseris** [comment]
merged!

**Comment by tkf** [comment]
Thanks!

**Comment by Datseris** [comment]
thank you!

**Pull request merged**